### ISSUE TITLE
saveInBackground/saveAsync not working with circular links

### ISSUE THREAD

**Issue by batkov** [ParsePlatform/Parse-SDK-iOS-OSX#587]
Hello,

I've tried this on Parse iOS SDK 1.10.0 / 1.9.1 / 1.8.5 / 1.8.0
Also local storage enabled. I think this is the key issue.

If we have circular links object not being saved. 
For example, we have following entities:
```
Message (text (string), like (Like))
Like (message (Message))
```

If we will try for existing Message create like, set links, so saveInBackground will stuck. No call to server performed.

```
- (void) setLikeToMessage:(Message) msg {
     Like * like = [Like object];
     like.message = msg;
     msg.like = like;
     [msg saveInBackgroundWithBlock:(BOOL, NSError*) {
            // Not called. Server call not performed
     }];
}
```
But this code will work:
```
- (void) setLikeToMessage:(Message) msg {
     Like * like = [Like object];
     like.message = msg;
     msg.like = like;
     [like saveInBackgroundWithBlock:(BOOL, NSError*) {
            // Called. Works properly
     }];
}
```

I did stop research on this issue, but then I found that I cannot sign up existing anonymous user.
I have following entities:
```
User (experience (Experience), personalProdile (PersonalProfile), ....another fields)
Experience (user (User), another fields)
PersonalProfile (user (User), another fields)
```

And for existing anonymous user Im' trying to do following:
```
+ (void)signUpWithUserName:(NSString *)userName 
password:(NSString *)password 
email:(NSString *)email
completion:(void(^)(NSError *error))completion {
    User *user = [User currentUser];
   // I did try to make [User objectWithoutDataWithObjectId:] but local data store get's data from DB.
    user.username = userName;
    user.password = password;
    user.email = email;
    
    [user signUpInBackgroundWithBlock:^(BOOL succeeded, NSError *error) {
        // Not called. Server call not performed
        if (completion) {
            completion(error);
        }
    }];
}
```

And it stuck for ever. I did debug and found that it goes to method called
```
+ (BFTask *)_deepSaveAsync:(id)object withCurrentUser:(PFUser *)currentUser sessionToken:(NSString *)sessionToken
```
And so **BFTask** created in this method never being called.

Please advise

**Comment by batkov** [comment]
@nlutsenko Why this issue is ignored? I can't start alpha testing of my app because of this issue.

**Comment by nlutsenko** [comment]
@batkov Sorry, it slipped through...
This is partly by design, but we should be explicit about it and i am not sure on why it doesn't throw.

You cannot save an object with a circular reference, if neither of the objects were saved to Parse.
This is due to the fact that we cannot create a pointer to non-existent object, when we are trying to save one or the other.

The solution here would be to save one object after another.

Let me know if this is not the case (any one of these objects have an objectId), otherwise I will look into why we don't surface this as an error.

**Comment by batkov** [comment]
Wait, wait... You are saying that sign Up not working is "by design"?
I'll post this part of code again:
```
User *user = [User currentUser];
   // I did try to make [User objectWithoutDataWithObjectId:] but local data store get's data from DB.
    user.username = userName;
    user.password = password;
    user.email = email;

    [user signUpInBackgroundWithBlock:^(BOOL succeeded, NSError *error) {
        // Not called. Server call not performed
        if (completion) {
            completion(error);
        }
    }];
```
Here I have User that was loginned anonymous. Then this user saved field "condition" with link to itself.
Then I'm trying to sign up with code above and no request performed, no callback called. And this by design?

I'll put this comment separately 
``` // I did try to make [User objectWithoutDataWithObjectId:] but local data store get's data from DB.
So there is no difference between calling [User currentUser] and [User objectWithoutDataWithObjectId: [User currentUser].objectId]

**Comment by nlutsenko** [comment]
This one is quite strange, as we should be able to convert the saved anonymous user to real user, even if there is an unsaved pointer back to it. Will look into the user signup issue separately. 

But overall (and if you never saved an anonymous user) - it is by design that you cannot save a new object with a new child that has a circular reference. 
Could you try signing up without a Profile and then attaching a profile and saving once again? This flow should definitely work.

**Comment by batkov** [comment]
We have Android app that works perfectly with same flow that I'm trying to implement on iOS.

I'm wondering why there is really "app stuck" if it is "by design". I'm sure developer should receive NSError "You need to save **some object** first."

> Could you try signing up without a Profile and then attaching a profile and saving once again? 
Sign Up new user works.

**Comment by nlutsenko** [comment]
Great, than it means that we indeed have an issue somewhere here.
I agree that returning an error here is the right flow and will look into why this doesn't do it (intended behavior is to fail with an error), as I already specified above.

P.S. A repro project with a small piece of code that pinpoints the issue would help tremendously.

**Comment by batkov** [comment]
@nlutsenko Sent you email with project

**Comment by nlutsenko** [comment]
Found a problem, it is indeed there, and it shouldn't fail/hang in your case.
Working on a fix, will send a pull request when it's ready.

**Comment by nlutsenko** [comment]
Hey @batkov, I've just attached a pull request to this issue that fixes the problem.
Since you are using CocoaPods - you can test it yourself by using this in your Podfile:
```ruby
pod 'Parse', :git => 'git@github.com:ParsePlatform/Parse-SDK-iOS-OSX.git', :branch => 'nlutsenko.user.save'
```


Let me know how it goes, but I've tested using your project and it signs the user up right now without hanging.

**Comment by batkov** [comment]
Permission denied (publickey).

fatal: Could not read from remote repository.
I'll wait for release. When it will be released?

**Comment by nlutsenko** [comment]
Will be released as part of `1.11.0` in the next few weeks.
